### PR TITLE
feature: add optional support for moved in / out blips

### PR DESCRIPTION
**Pull request by setchy**
Hot on the heels of the [v1.0.0](https://github.com/thoughtworks/build-your-own-radar/releases/tag/v1.0.0) launch this week, I thought I'd see whether I could self-serve #177 

Adds optional support for moved in and out blip drawing.

Below are two datasets for Thoughtworks Volume 27 sample (csv) that can be used to demonstrate/test:
* [Thoughtworks Volume 27 with movements](https://raw.githubusercontent.com/setchy/thoughtworks-tech-radar-volumes/main/volumes/Thoughtworks%20Technology%20Radar%20Volume%2027.csv)
* [Thoughtworks Volume 27 without movements](https://docs.google.com/spreadsheets/d/1sL6j3eYxsiXMtgELi52pn6XMzk_L4ZtfxYr1KjNYZpo/edit#gid=0)

Screenshot 2023-04-22 at 4 18 08 PM

### CODE REVIEW

**Comment by setchy** [comment]
@devansh-sharma-tw - appreciate your feedback re this feature proposal

**Comment by setchy** [comment]
@marisahoenig - would you be able take this to the team for review please

**Comment by devansh-sharma-tw** [comment]
Thanks for raising this @setchy ! We're internally discussing on adding the moved in/out status as well (to keep BYOR inline with our Thoughtworks Radar).
This PR helps a lot for this and we'll update here soon regarding the feature.
Thanks!

**Comment by setchy** [comment]
@devansh-sharma-tw - hope all is well. Just checking in to see if your team has had a chance to review and consider this PR.

File: spec/util/inputSanitizer-spec.js [hunk 1]
```diff
@@ -107,6 +107,8 @@ describe('Input Santizer for Protected sheet', function () {
       ring: '',
       quadrant: '',
       isNew: '',
+      hasMovedIn: '',
```
[the remark this answers is not in the document]
Comment (marisahoenig) [reply spec/util/inputSanitizer-spec.js:110]: Thinking about this more, I like the idea of having one field that can be utilized instead of `isNew` since a blip can be new OR moved in OR moved out OR a refresh.

**Review by marisahoenig — CHANGES_REQUESTED** [review]
Adding some suggestions. I think it's worth adding one new field and letting users have the option to use it OR the `isNew` field. Would love your thoughts @setchy

File: spec/util/inputSanitizer-spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (setchy) [reply spec/util/inputSanitizer-spec.js:110]: I agree, a new field `status` or similar  

Would you be willing to make this a breaking change, or want to support backwards compatibility with datasets that are using `isNew`? 

Can you elaborate on your definition of `refresh`.  Is this for blips that are removed from the radar and re-appear, or something different?

File: src/graphing/blips.js [hunk 2]
```diff
@@ -104,9 +104,20 @@ function findBlipCoordinates(blip, minRadius, maxRadius, startAngle, allBlipCoor
...
+  } else if (blip.hasMovedOut()) {
+    blipType = 'moved out'
+  } else {
+    blipType = 'existing'
+  }
+
   return blip.isGroup()
     ? `\`${blip.ring().name()} ring, group of ${blip.blipText()}`
-    : `${blip.ring().name()} ring, ${blip.name()}, ${blip.isNew() ? 'new' : 'existing'} blip.`
+    : `${blip.ring().name()} ring, ${blip.name()}, ${blipType} blip.`
```
[the remark this answers is not in the document]
Comment (setchy) [reply src/graphing/blips.js:120]: Agree :)

File: src/stylesheets/_quadrants.scss [hunk 3]
```diff
@@ -402,8 +402,8 @@
       top: calc($quadrantWidth * 2 + $quadrantsGap);
     }
 
-    img:nth-child(2) {
-      margin-left: 48px;
+    img:nth-child(n + 2) {
```
[the remark this answers is not in the document]
Comment (setchy) [reply src/stylesheets/_quadrants.scss:405]: I don't off the top of my head, it's been too long...

File: spec/util/inputSanitizer-spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (marisahoenig) [reply spec/util/inputSanitizer-spec.js:110]: I'd prefer making it backward compatible. So, users should have the option to _either_ use `isNew` or `status`. If there's a value in `status`, we would use that field.

And here's where the differences between TW Tech Radar and BYOR come back 😊 The "refresh" I'm referring to is what shows as "No change" on https://www.thoughtworks.com/radar. Basically, the blip has appeared before and we're blipping it again, but it stays in the same ring as last time. In BYOR, this would be a `false` for `isNew`.

File: src/stylesheets/_quadrants.scss (hunk 3 above)
[the remark this answers is not in the document]
Comment (marisahoenig) [reply src/stylesheets/_quadrants.scss:405]: Fair! All good then.

File: spec/util/inputSanitizer-spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (setchy) [reply spec/util/inputSanitizer-spec.js:110]: Ok great, that clears things up.  Would you want to keep it as `refresh`, or prefer something different?

[the remark this answers is not in the document]
Comment (marisahoenig) [reply spec/util/inputSanitizer-spec.js:110]: Since `refresh` isn't too intuitive, I'm open to suggestions! I was thinking maybe `existing` but not sure on that. `notNew` seems not great too.

[the remark this answers is not in the document]
Comment (setchy) [reply spec/util/inputSanitizer-spec.js:110]: I think either `existing` or `noChange` would work. Whichever we use should match the terminology used in the legend.

File: src/stylesheets/_quadrants.scss (hunk 3 above)
[the remark this answers is not in the document]
Comment (setchy) [reply src/stylesheets/_quadrants.scss:405]: Ah, i think I recall, it was to support the legend going from 2 items to 3 items

**Comment by setchy** [comment]
@marisahoenig - I have updated this PR to support a new, optional column/field called `status`.  

Accepted values are case-insensitive `new`, `no change`, `moved in`, `moved out`.  I welcome any feedback on alternate option names (eg: `existing` or `unchanged` etc)

To demonstrate, I've updated my [TW volume datasets](https://github.com/setchy/thoughtworks-tech-radar-volumes/tree/main/volumes), eg [TW Volume 28](https://github.com/setchy/thoughtworks-tech-radar-volumes/blob/main/volumes/Thoughtworks%20Technology%20Radar%20Volume%2028.csv).

All existing datasets using `isNew` should continue to work.

**Comment by setchy** [comment]
> [!NOTE]
> in the interim I've deployed my BYOR enhancements @ https://radar.setchy.io, which bring it up to the functionality found on https://thoughtworks.com/radar

**Comment by danielkoch** [comment]
@marisahoenig: Any plans or update on this one?

File: spec/graphing/blips-spec.js [hunk 4]
```diff
@@ -187,12 +187,89 @@ describe('Blips', function () {
...
+
+  it('should return correct assistive text for existing blip', function () {
+    const blip = {
+      isGroup: () => false,
+      ring: () => {
+        return {
+          name: () => 'ring1',
+        }
+      },
+      blipText: () => '12 New Blips',
```
Comment (marisahoenig) [inline spec/graphing/blips-spec.js:224]: For this test, it should be existing blips — also the `actual` statement should be something like "group of 12 Existing Blips"

Comment (marisahoenig) [reply spec/graphing/blips-spec.js:224]: Should there be a status() value here?

File: spec/graphing/blips-spec.js [hunk 5]
```diff
@@ -187,12 +187,89 @@ describe('Blips', function () {
...
+      name: () => 'blip1',
+      isNew: () => false,
+      hasMovedIn: () => true,
+      hasMovedOut: () => false,
+    }
+
+    const actual = blipAssistiveText(blip)
+    expect(actual).toEqual('ring1 ring, blip1, moved in blip.')
```
Comment (marisahoenig) [inline spec/graphing/blips-spec.js:251]: `actual` here needs to be revised too. And I think status() should be added as well?

File: spec/graphing/blips-spec.js [hunk 6]
```diff
@@ -187,12 +187,89 @@ describe('Blips', function () {
...
+      },
+      blipText: () => '12 New Blips',
+      name: () => 'blip1',
+      isNew: () => false,
+      hasMovedIn: () => false,
+      hasMovedOut: () => true,
+    }
+
+    const actual = blipAssistiveText(blip)
+    expect(actual).toEqual('ring1 ring, blip1, moved out blip.')
```
Comment (marisahoenig) [inline spec/graphing/blips-spec.js:270]: Same comment here

File: src/graphing/radar.js [hunk 7]
```diff
@@ -828,12 +828,30 @@ const Radar = function (size, radar) {
...
 
+  function hasMovementData(quadrants) {
+    let hasMovements = false
+
+    for (var quadrantWrapper of quadrants) {
+      let quadrant = quadrantWrapper.quadrant
+
+      for (var blip of quadrant.blips()) {
+        if (blip.status() !== '') {
+          hasMovements = true
```
Comment (marisahoenig) [inline src/graphing/radar.js:846]: Could we just `return true` here instead (and then `return false` instead of `return hasMovements`) to get rid of the `hasMovements` variable?

File: src/graphing/radar.js [hunk 8]
```diff
@@ -828,12 +828,30 @@ const Radar = function (size, radar) {
     })
 
     if (featureToggles.UIRefresh2022) {
-      renderRadarLegends(radarElement)
+      let hasMovements = hasMovementData(quadrants)
+      renderRadarLegends(radarElement, hasMovements)
```
Comment (marisahoenig) [inline src/graphing/radar.js:832]: Consolidate into:
`renderRadarLegends(radarElement, hasMovementData(quadrants))` since we don't use the `hasMovements` var again

File: spec/models/blip-spec.js [hunk 9]
```diff
@@ -67,6 +67,30 @@ describe('Blip', function () {
     expect(blip.isNew()).toBe(false)
   })
 
+  it('status is new', function () {
```
Comment (marisahoenig) [inline spec/models/blip-spec.js:70]: Can you add topic and description to these tests so it's easier to follow which params are which?

File: src/graphing/blips.js [hunk 10]
```diff
@@ -104,9 +104,21 @@ function findBlipCoordinates(blip, minRadius, maxRadius, startAngle, allBlipCoor
...
 function blipAssistiveText(blip) {
+  let blipType
+
+  if (blip.isNew()) {
+    blipType = 'new'
+  } else if (blip.hasMovedIn()) {
+    blipType = 'moved in'
+  } else if (blip.hasMovedOut()) {
+    blipType = 'moved out'
+  } else {
```
Comment (marisahoenig) [inline src/graphing/blips.js:115]: Should the `hasNoChange` fxn be incorporated here?

File: src/graphing/blips.js [hunk 11]
```diff
@@ -104,9 +104,21 @@ function findBlipCoordinates(blip, minRadius, maxRadius, startAngle, allBlipCoor
...
+    blipType = 'existing'
+  }
+
   return blip.isGroup()
     ? `\`${blip.ring().name()} ring, group of ${blip.blipText()}`
-    : `${blip.ring().name()} ring, ${blip.name()}, ${blip.isNew() ? 'new' : 'existing'} blip.`
+    : `${blip.ring().name()} ring, ${blip.name()}, ${blipType} blip.`
```
Comment (marisahoenig) [inline src/graphing/blips.js:121]: Very nice!

**Review by marisahoenig — CHANGES_REQUESTED** [review]
Hey @setchy I left a few comments. Please resolve and also make sure the most recent code from the main branch of this repo is loaded in. I'll chat with the team next week to see if we can get this added. The slowdown was related to the most recent Radar release, and we need to make sure changes like this are reflected on our live site as well. Apologies for the delay — it's a great add to the tool.

File: spec/graphing/blips-spec.js (hunk 5 above)
Comment (setchy) [reply spec/graphing/blips-spec.js]: Ah yes, great catch.  Upon inspecting thoughtworks.com/radar I can see that the assistive text format has been updated/fixed in the past 12mths.  

I'll pushed an update which aligns this with the following observed format
`<ring name> ring, <blip name>, <New|Moved in|Moved out|No change>.`

File: spec/graphing/blips-spec.js (hunk 4 above)
Comment (setchy) [reply spec/graphing/blips-spec.js]: Since this test case wasn't for grouped blips, I've removed it

File: spec/graphing/blips-spec.js (hunk 6 above)
Comment (setchy) [reply spec/graphing/blips-spec.js]: Since this test case wasn't for grouped blips, I've removed it

File: src/graphing/radar.js (hunk 7 above)
Comment (setchy) [reply src/graphing/radar.js]: Deal 🤝

File: src/graphing/blips.js (hunk 10 above)
Comment (setchy) [reply src/graphing/blips.js]: Simplified to use blip.status directly (previously the assistive text was formatted in a way that wasn't compatible.  Now, not the case :)

File: src/graphing/radar.js (hunk 8 above)
Comment (setchy) [reply src/graphing/radar.js]: Deal 🤝

File: spec/models/blip-spec.js (hunk 9 above)
Comment (setchy) [reply spec/models/blip-spec.js:70]: Sorted.

**Comment by setchy** [comment]
> Hey @setchy I left a few comments. Please resolve and also make sure the most recent code from the main branch of this repo is loaded in. I'll chat with the team next week to see if we can get this added. The slowdown was related to the most recent Radar release, and we need to make sure changes like this are reflected on our live site as well. Apologies for the delay — it's a great add to the tool.

Thanks @marisahoenig, appreciate the feedback.  I've gone through and addressed each of them 😄.  

No problems about the delay, too - completely understand.

**Review by marisahoenig — APPROVED** [review]
Thanks for all the changes, @setchy! As mentioned, I'll chat with the team next week about getting this incorporated and seeing if there is anything else that needs to be updated on the team's end.

**Comment by setchy** [comment]
@will-amaral - appreciate your thoughts on this PR

**Review by shiviraj — APPROVED** [review]
